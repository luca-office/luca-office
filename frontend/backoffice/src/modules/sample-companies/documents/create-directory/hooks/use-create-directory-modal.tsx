import {useForm, UseFormMethods} from "react-hook-form"
import {Directory} from "shared/models"
import {Option} from "shared/utils"
import {DirectoryCreation} from "../../../../../graphql/generated/globalTypes"
import {useCreateSampleCompanyDirectory} from "../../../../../graphql/hooks/mutations/directory/create-directory"
import {DirectoryCreationForm} from "../../../../common/files-and-directories/create-directory/create-directory-modal"

export interface UseCreateDirectoryHook {
  readonly createDirectory: (creation: DirectoryCreation) => Promise<Option<Directory>>
  readonly createDirectoryLoading: boolean
  readonly formMethods: UseFormMethods<DirectoryCreationForm>
}

export const useCreateDirectoryModal = (sampleCompanyId: UUID): UseCreateDirectoryHook => {
  const {createDirectory, createDirectoryLoading} = useCreateSampleCompanyDirectory(sampleCompanyId)
  const formMethods = useForm<DirectoryCreationForm>()

  return {
    createDirectory,
    createDirectoryLoading,
    formMethods
  }
}
