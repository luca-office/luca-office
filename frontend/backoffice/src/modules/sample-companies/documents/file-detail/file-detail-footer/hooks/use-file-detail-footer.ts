import {useState} from "react"
import {useDispatch} from "react-redux"
import {Relevance} from "shared/graphql/generated/globalTypes"
import {
  useCreateScenarioSampleCompanyFile,
  useInterventions,
  useScenario,
  useScenarioSampleCompanyFiles,
  useUpdateScenarioSampleCompanyFile
} from "shared/graphql/hooks"
import {FileOpeningIntervention, ScenarioSampleCompanyFile, SelectOption} from "shared/models"
import {useLucaTranslation} from "shared/translations"
import {find, first, getRelevanceCriteria, Option} from "shared/utils"
import {navigateToRouteAction} from "../../../../../../redux/actions/navigation-action"
import {Route} from "../../../../../../routes"
import {getInterventionsForTypeName, interventionTypeToGroupType} from "../../../../../scenario-interventions/utils"

export interface UseFileDetailFooterHook {
  readonly interventionCount: number
  readonly isCreateInterventionModalVisible: boolean
  readonly navigateToIntervention: () => void
  readonly onCreateInterventionClick: () => void
  readonly relevanceOptions: SelectOption[]
  readonly scenarioSampleCompanyFile: Option<ScenarioSampleCompanyFile>
  readonly updateRelevance: (relevance: Relevance) => void
  readonly toggleIsCreateInterventionModalVisible: () => void
  readonly scenarioDurationInSeconds: number
}

export const useFileDetailFooter = (scenarioId: UUID, fileId: UUID): UseFileDetailFooterHook => {
  const {t} = useLucaTranslation()

  const [isCreateInterventionModalVisible, setIsCreateInterventionModalVisible] = useState(false)

  const {scenarioSampleCompanyFiles} = useScenarioSampleCompanyFiles(scenarioId)
  const {scenario} = useScenario(scenarioId)
  const {createScenarioSampleCompanyFile} = useCreateScenarioSampleCompanyFile()
  const {updateScenarioSampleCompanyFile} = useUpdateScenarioSampleCompanyFile()
  const {interventions} = useInterventions(scenarioId)

  const dispatch = useDispatch()

  const scenarioSampleCompanyFile = find(entry => entry.fileId === fileId, scenarioSampleCompanyFiles)

  const interventionsForFile = (getInterventionsForTypeName(
    interventions.getOrElse([]),
    "FileOpeningIntervention"
  ) as FileOpeningIntervention[]).filter(intervention => intervention.fileId === fileId)

  const relevanceOptions = getRelevanceCriteria(t)

  const updateRelevance = (relevance: Relevance) => {
    if (scenarioSampleCompanyFile.isEmpty()) {
      createScenarioSampleCompanyFile({scenarioId, fileId, relevance})
      return
    }
    updateScenarioSampleCompanyFile(scenarioId, fileId, {relevance})
  }

  const navigateToIntervention = () =>
    first(interventionsForFile).forEach(intervention => {
      const interventionGroupType = interventionTypeToGroupType(intervention.interventionType)

      dispatch(
        navigateToRouteAction(Route.ScenarioInterventionsGroupEntityDetail, {
          scenarioId,
          groupType: interventionGroupType,
          headerGroupType: interventionGroupType,
          groupEntityId: fileId
        })
      )
    })

  return {
    scenarioSampleCompanyFile,
    scenarioDurationInSeconds: scenario.map(s => s.maxDurationInSeconds || 0).getOrElse(0),
    relevanceOptions,
    updateRelevance,
    navigateToIntervention,
    interventionCount: interventionsForFile.length,
    onCreateInterventionClick: () => setIsCreateInterventionModalVisible(true),
    isCreateInterventionModalVisible,
    toggleIsCreateInterventionModalVisible: () => setIsCreateInterventionModalVisible(!isCreateInterventionModalVisible)
  }
}
