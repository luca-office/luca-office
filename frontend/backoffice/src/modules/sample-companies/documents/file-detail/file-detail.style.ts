import {css} from "@emotion/react"
import {
  borderRadius,
  spacing,
  spacingHuge,
  spacingLarge,
  spacingMedium,
  spacingSmall,
  spacingTiny,
  TextSize
} from "shared/styles"

export const fileDetailStyles = {
  wrapper: css({
    display: "flex",
    flexDirection: "column",
    height: `calc(100% - ${2 * spacingLarge}px)`
  }),
  subject: css({
    margin: spacing(0, -spacingSmall),

    "form > div": {
      padding: spacing(spacingTiny, spacingSmall, spacingTiny, spacingSmall)
    },
    input: {
      fontSize: TextSize.Larger
    }
  }),
  viewerTool: css({
    height: "80vh",
    width: "80vh"
  }),
  viewerToolPlaceholder: css({
    padding: spacingLarge,
    backgroundColor: "white",
    borderRadius: borderRadius
  }),
  content: css({
    padding: spacingHuge,
    width: "initial",
    justifyContent: "flex-start"
  }),
  nameAndDirectoryWrapper: css({
    display: "grid",
    gridTemplateColumns: "1fr minmax(0, 1fr)",
    gap: spacingMedium
  }),
  actionField: css({
    width: "100%",

    ".content": {
      minWidth: 0
    }
  }),
  textHint: css({
    margin: spacing(-spacingMedium, 0, spacingSmall)
  })
}
