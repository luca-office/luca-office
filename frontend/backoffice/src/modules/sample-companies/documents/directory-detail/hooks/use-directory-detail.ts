import {Dispatch, SetStateAction, useState} from "react"
import {useDispatch} from "react-redux"
import {FileUsageType, Relevance, TextDocumentCreation} from "shared/graphql/generated/globalTypes"
import {useCreateSampleCompanyFile, useFilesForSampleCompany, useSampleCompany} from "shared/graphql/hooks"
import {useCreateTextDocument} from "shared/graphql/hooks/mutations/text-document/use-create-textdocument"
import {File, SampleCompany, UploadBinary} from "shared/models"
import {TextDocumentCreationInfo} from "shared/models/text-document-creation-info"
import {find, isDefined, Option} from "shared/utils"
import {DirectoryUpdate} from "../../../../../graphql/generated/globalTypes"
import {useDirectoriesForSampleCompany} from "../../../../../graphql/hooks"
import {useUpdateDirectory} from "../../../../../graphql/hooks/mutations/directory/update-directory"
import {Directory} from "../../../../../models"
import {navigateToRouteAction} from "../../../../../redux/actions/navigation-action"
import {expandDirectoriesAction} from "../../../../../redux/actions/ui/scenarios/directories-and-files-action"
import {Route} from "../../../../../routes"

export interface UseDirectoryDetailHook {
  readonly updateInProgress: boolean
  readonly updateDirectory: (id: UUID, update: DirectoryUpdate) => Promise<Option<Directory>>
  readonly selectDirectory: (directoryId: UUID) => void
  readonly selectFile: (fileId: UUID) => void
  readonly deselectDirectory: () => void
  readonly isCreateFileOverlayVisible: boolean
  readonly isCreateSubdirectoryOverlayVisible: boolean
  readonly showCreateFileOverlay: () => void
  readonly hideCreateFileOverlay: () => void
  readonly showCreateSubdirectoryOverlay: () => void
  readonly hideCreateSubdirectoryOverlay: () => void
  readonly handleCreateSubdirectory: (directoryId: UUID, subdirectoryId: UUID) => void
  readonly createFilesFromBinaries: (uploadedBinaries: UploadBinary[], directoryId: UUID) => void
  readonly isUpdateDirectoryOverlayVisible: boolean
  readonly setIsUpdateDirectoryOverlayVisible: Dispatch<SetStateAction<boolean>>
  readonly allDirectories: Directory[]
  readonly allFiles: File[]
  readonly dataLoading: boolean
  readonly sampleCompany: Option<SampleCompany>
  readonly isPublished: boolean
  readonly createTextDocumentFile: (textDocumentCreationInfo: TextDocumentCreationInfo) => void
}

export const useDirectoryDetail = (sampleCompanyId: UUID, scenarioId?: UUID): UseDirectoryDetailHook => {
  const dispatch = useDispatch()

  const {sampleCompany, sampleCompanyLoading} = useSampleCompany(sampleCompanyId)
  const {updateDirectory, updateDirectoryLoading} = useUpdateDirectory()
  const {createSampleCompanyFile: createFile} = useCreateSampleCompanyFile(sampleCompanyId)
  const {directories, directoriesLoading} = useDirectoriesForSampleCompany(sampleCompanyId)
  const {files, filesLoading} = useFilesForSampleCompany(sampleCompanyId)
  const {createTextDocument} = useCreateTextDocument({sampleCompanyId})

  const [isCreateFileOverlayVisible, updateIsCreateFileOverlayVisible] = useState(false)
  const [isCreateSubdirectoryOverlayVisible, updateIsCreateSubdirectoryOverlayVisible] = useState(false)
  const [isUpdateDirectoryOverlayVisible, setIsUpdateDirectoryOverlayVisible] = useState(false)

  const isPublished = sampleCompany.exists(({publishedAt}) => isDefined(publishedAt))

  const showCreateFileOverlay = () => updateIsCreateFileOverlayVisible(true)

  const hideCreateFileOverlay = () => updateIsCreateFileOverlayVisible(false)

  const showCreateSubdirectoryOverlay = () => updateIsCreateSubdirectoryOverlayVisible(true)

  const hideCreateSubdirectoryOverlay = () => updateIsCreateSubdirectoryOverlayVisible(false)

  const handleUpdateDirectory = (id: UUID, update: DirectoryUpdate) => updateDirectory(id, update)

  const navigateToDirectory = (directoryId: UUID) => {
    dispatch(
      navigateToRouteAction(
        scenarioId ? Route.ScenarioAssignedSampleCompanyDocumentsDetails : Route.SampleCompanyDocumentDetails,
        {scenarioId, sampleCompanyId, directoryId}
      )
    )
  }

  const handleCreateSubdirectory = (directoryId: UUID, subdirectoryId: UUID) => {
    hideCreateSubdirectoryOverlay()
    navigateToDirectory(subdirectoryId)
    dispatch(expandDirectoriesAction([directoryId]))
  }

  const createFilesFromBinaries = (uploadedBinaries: UploadBinary[], directoryId: UUID) => {
    Promise.all(
      uploadedBinaries.map(binary =>
        createFile({
          name: binary.data.filename,
          relevance: Relevance.PotentiallyHelpful,
          tags: [],
          usageType: FileUsageType.FileSystem,
          directoryId,
          ...{[binary.isSpreadsheet ? "spreadsheetId" : "binaryFileId"]: binary.data.id}
        })
      )
    )
      .then(hideCreateFileOverlay)
      .catch(err => console.error(err))
  }

  const selectDirectory = (directoryId: UUID) => navigateToDirectory(directoryId)

  const selectFile = (fileId: UUID) =>
    find(file => file.id === fileId, files.getOrElse([])).forEach(file =>
      dispatch(
        navigateToRouteAction(
          scenarioId ? Route.ScenarioAssignedSampleCompanyDocumentsDetailsFile : Route.SampleCompanyDocumentDetailsFile,
          {
            scenarioId,
            sampleCompanyId,
            directoryId: file.directoryId,
            fileId: file.id
          }
        )
      )
    )

  const deselectDirectory = () =>
    dispatch(
      navigateToRouteAction(scenarioId ? Route.ScenarioAssignedSampleCompanyDocuments : Route.SampleCompanyDocuments, {
        scenarioId,
        sampleCompanyId
      })
    )

  const createTextDocumentFile = (textDocumentCreationInfo: TextDocumentCreationInfo) => {
    createTextDocument(textDocumentCreationInfo.textDocument).then(textDocumentOption =>
      textDocumentOption.forEach(textDocument =>
        createFile({
          name: textDocumentCreationInfo.title,
          relevance: Relevance.PotentiallyHelpful,
          tags: [],
          usageType: FileUsageType.FileSystem,
          textDocumentId: textDocument.id,
          directoryId: textDocumentCreationInfo.directoryId
        })
      )
    )
  }

  return {
    updateInProgress: updateDirectoryLoading,
    updateDirectory: handleUpdateDirectory,
    selectDirectory,
    selectFile,
    deselectDirectory,
    isCreateFileOverlayVisible,
    showCreateFileOverlay,
    hideCreateFileOverlay,
    isCreateSubdirectoryOverlayVisible,
    showCreateSubdirectoryOverlay,
    hideCreateSubdirectoryOverlay,
    createFilesFromBinaries,
    handleCreateSubdirectory,
    isUpdateDirectoryOverlayVisible,
    setIsUpdateDirectoryOverlayVisible,
    allDirectories: directories.getOrElse([]),
    allFiles: files.getOrElse([]),
    dataLoading: directoriesLoading || filesLoading || sampleCompanyLoading,
    sampleCompany,
    isPublished,
    createTextDocumentFile
  }
}
