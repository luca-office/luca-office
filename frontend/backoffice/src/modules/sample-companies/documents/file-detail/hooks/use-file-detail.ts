import * as React from "react"
import {useDispatch} from "react-redux"
import {FileType} from "shared/enums"
import {FileUpdate} from "shared/graphql/generated/globalTypes"
import {useFilesForSampleCompany, useSampleCompany, useScenario, useUpdateFile} from "shared/graphql/hooks"
import {File} from "shared/models"
import {addFileExtension, fileExtensionForFile, isDefined, Option} from "shared/utils"
import {useDirectoriesForSampleCompany} from "../../../../../graphql/hooks"
import {Directory} from "../../../../../models"
import {navigateToRouteAction} from "../../../../../redux/actions/navigation-action"
import {Route} from "../../../../../routes"
import {PreviewInfo} from "../../../../common/files-and-directories/file-detail/file-detail-utils"

export interface UseFileDetailHook {
  readonly updateFile: (update: Partial<FileUpdate>, file: File) => Promise<Option<File>>
  readonly deselectFile: () => void
  readonly filePreviewOption: Option<PreviewInfo>
  readonly setFilePreviewOption: React.Dispatch<React.SetStateAction<Option<PreviewInfo>>>
  readonly isMoveOverlayVisible: boolean
  readonly showMoveOverlay: () => void
  readonly hideMoveOverlay: () => void
  readonly allDirectories: Directory[]
  readonly allFiles: File[]
  readonly dataLoading: boolean
  readonly isSampleCompanyPublished: boolean
  readonly isScenarioPublishedOrFinalized: boolean
}

export const useFileDetail = (sampleCompanyId: UUID, scenarioId?: UUID): UseFileDetailHook => {
  const dispatch = useDispatch()

  const [filePreviewOption, setFilePreviewOption] = React.useState<Option<PreviewInfo>>(Option.none())
  const [isMoveOverlayVisible, setIsMoveOverlayVisible] = React.useState<boolean>(false)

  const {sampleCompany, sampleCompanyLoading} = useSampleCompany(sampleCompanyId)
  const {updateFile, updateFileLoading} = useUpdateFile()
  const {scenario, scenarioLoading} = useScenario(scenarioId || "", scenarioId === undefined)
  const {directories, directoriesLoading} = useDirectoriesForSampleCompany(sampleCompanyId)
  const {files, filesLoading} = useFilesForSampleCompany(sampleCompanyId)

  const allDirectories = directories
    .getOrElse([])
    .map(directory =>
      sampleCompany
        .map(({directoryId, name}) =>
          directory.id === directoryId ? {...directory, name: directory.name || name} : directory
        )
        .getOrElse(directory)
    )

  const isPublished = sampleCompany.exists(({publishedAt}) => isDefined(publishedAt))

  const deselectFile = () =>
    dispatch(
      navigateToRouteAction(
        scenarioId ? Route.ScenarioAssignedSampleCompanyDocumentsDetails : Route.SampleCompanyDocumentDetails,
        {
          scenarioId,
          sampleCompanyId,
          directoryId: sampleCompany.map(company => company.directoryId).orUndefined()
        }
      )
    )

  const showMoveOverlay = () => {
    setIsMoveOverlayVisible(true)
  }
  const hideMoveOverlay = () => {
    setIsMoveOverlayVisible(false)
  }

  const handleUpdateFile = (update: Partial<FileUpdate>, file: File) => {
    if (updateFileLoading) {
      return Promise.reject()
    }

    const binaryFileId =
      file.fileType === FileType.Media
        ? update.binaryFileId !== undefined
          ? update.binaryFileId
          : file.binaryFileId
        : null

    return updateFile(file.id, {
      name: update.name !== undefined ? addFileExtension(update.name, fileExtensionForFile(file)) : file.name,
      binaryFileId,
      relevance: update.relevance !== undefined ? update.relevance : file.relevance,
      tags: update.tags !== undefined ? update.tags : file.tags,
      directoryId: update.directoryId !== undefined ? update.directoryId : file.directoryId
    })
  }

  return {
    updateFile: handleUpdateFile,
    deselectFile,
    filePreviewOption,
    setFilePreviewOption,
    isMoveOverlayVisible,
    showMoveOverlay,
    hideMoveOverlay,
    allDirectories,
    allFiles: files.getOrElse([]),
    dataLoading: directoriesLoading || filesLoading || sampleCompanyLoading || scenarioLoading,
    isSampleCompanyPublished: isPublished,
    isScenarioPublishedOrFinalized: scenario.exists(s => s.publishedAt !== null || s.finalizedAt !== null)
  }
}
