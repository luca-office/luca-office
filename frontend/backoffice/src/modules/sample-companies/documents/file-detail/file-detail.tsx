import {CSSInterpolation} from "@emotion/serialize"
import * as React from "react"
import {
  Card,
  CardContent,
  FormFieldLabel,
  MediaFilePreview,
  Overlay,
  ReadonlyActionField,
  Tooltip
} from "shared/components"
import {ViewerToolsType} from "shared/enums"
import {useDeleteFileFromSampleCompany} from "shared/graphql/hooks"
import {DirectoryNode, File} from "shared/models"
import {CustomStyle} from "shared/styles"
import {useLucaTranslation} from "shared/translations"
import {iconForFile, Option, removeFileExtension} from "shared/utils"
import {InlineEditableHeaderContainer} from "../../../../components"
import {HintText} from "../../../../components/hint-text/hint-text"
import {DetailHeader} from "../../../common/files-and-directories/file-detail/detail-header/detail-header"
import {PreviewInfo, renderViewerToolForFile} from "../../../common/files-and-directories/file-detail/file-detail-utils"
import {PathActionField} from "../../../common/files-and-directories/file-detail/path-action-field/path-action-field"
import {MoveOverlay} from "../../../common/files-and-directories/move/move-overlay"
import {fileDetailStyles as styles} from "./file-detail.style"
import {FileDetailFooter} from "./file-detail-footer/file-detail-footer"
import {useFileDetail} from "./hooks/use-file-detail"

export interface FileDetailProps extends CustomStyle {
  readonly customContentStyles?: CSSInterpolation
  readonly disabled: boolean
  readonly isIntroFile: boolean
  readonly isLogoFile: boolean
  readonly file: File
  readonly parentDirectory: Option<DirectoryNode>
  readonly sampleCompanyId: UUID
  readonly scenarioId?: UUID
}

export const FileDetail: React.FC<FileDetailProps> = ({
  customContentStyles,
  customStyles,
  disabled,
  isIntroFile,
  isLogoFile,
  file,
  parentDirectory,
  sampleCompanyId,
  scenarioId
}) => {
  const {t} = useLucaTranslation()

  const {
    updateFile,
    deselectFile,
    filePreviewOption,
    setFilePreviewOption,
    isMoveOverlayVisible,
    showMoveOverlay,
    hideMoveOverlay,
    allDirectories,
    allFiles,
    dataLoading,
    isSampleCompanyPublished,
    isScenarioPublishedOrFinalized
  } = useFileDetail(sampleCompanyId, scenarioId)

  const closeViewerTool = () => setFilePreviewOption(Option.none())

  const closeBinary = () => setFilePreviewOption(Option.none())

  const onPreviewButtonClick = (viewerTool: ViewerToolsType) =>
    setFilePreviewOption(Option.of<PreviewInfo>({file, viewerTool}))

  return (
    <React.Fragment>
      <Card customStyles={customStyles}>
        <DetailHeader
          disabled={disabled || isSampleCompanyPublished}
          deleteHook={() => useDeleteFileFromSampleCompany(sampleCompanyId)}
          onDelete={deselectFile}
          entityId={file.id}
          title={file.name}
          icon={iconForFile(file)}
        />
        <div css={[styles.wrapper, customContentStyles]}>
          <CardContent customStyles={styles.content}>
            {(isIntroFile || isLogoFile) && (
              <HintText
                text={isIntroFile ? t("sample_company_details__intro_hint") : t("sample_company_details__logo_hint")}
                customStyles={styles.textHint}
              />
            )}
            <div css={styles.nameAndDirectoryWrapper}>
              <div>
                <FormFieldLabel textKey="files_and_directories__file_detail_name" />
                <Tooltip
                  title={t("sample_companies_files__disabled_tooltip")}
                  inactive={!disabled && !isSampleCompanyPublished}>
                  <InlineEditableHeaderContainer
                    onConfirm={name => updateFile({name}, file)}
                    customStyles={styles.subject}
                    disabled={disabled || isSampleCompanyPublished}
                    text={removeFileExtension(file.name)}
                  />
                </Tooltip>
              </div>
              <div>
                <FormFieldLabel textKey="files_and_directories__move_directory_label" />
                <ReadonlyActionField
                  customStyles={styles.actionField}
                  renderValue={() => (
                    <PathActionField
                      disabled={disabled || isSampleCompanyPublished}
                      tooltipLabel={t("sample_companies_files__disabled_tooltip")}
                      onMoveClick={showMoveOverlay}
                      t={t}
                      parentDirectory={parentDirectory}
                    />
                  )}
                  disabled={disabled || isSampleCompanyPublished}
                />
              </div>
            </div>
            <MediaFilePreview onPreviewButtonClick={onPreviewButtonClick} file={file} />
          </CardContent>
          {!!scenarioId && (
            <FileDetailFooter
              isScenarioPublished={isScenarioPublishedOrFinalized}
              scenarioId={scenarioId}
              file={file}
              disabled={disabled}
            />
          )}
        </div>
      </Card>
      {filePreviewOption
        .map(filePreview => (
          <Overlay>{renderViewerToolForFile(filePreview, closeViewerTool, closeBinary, t, sampleCompanyId)}</Overlay>
        ))
        .orNull()}
      {isMoveOverlayVisible && (
        <MoveOverlay
          file={Option.of(file)}
          directory={Option.none()}
          onFilesSuccessfullyMoved={hideMoveOverlay}
          onDismiss={hideMoveOverlay}
          name={file.name}
          parentDirectory={parentDirectory}
          dataLoading={dataLoading}
          directories={allDirectories}
          files={allFiles}
        />
      )}
    </React.Fragment>
  )
}
