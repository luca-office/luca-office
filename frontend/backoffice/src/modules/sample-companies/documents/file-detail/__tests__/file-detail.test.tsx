import {MockedProvider} from "@apollo/client/testing"
import {mount} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {
  Card,
  CardContent,
  FormFieldLabel,
  ImageViewer,
  MediaFilePreview,
  Overlay,
  PdfViewer,
  Text
} from "shared/components"
import {ViewerToolsType} from "shared/enums"
import {fileMock, filesMock, sampleCompaniesMock} from "shared/graphql/__mocks__"
import {deleteFileMutation} from "shared/graphql/mutations"
import {Option} from "shared/utils"
import {InlineEditableHeaderContainer} from "../../../../../components"
import {HintText} from "../../../../../components/hint-text/hint-text"
import {directoriesForScenarioMock} from "../../../../../graphql/__mocks__"
import {DetailHeader} from "../../../../common/files-and-directories/file-detail/detail-header/detail-header"
import {PreviewInfo} from "../../../../common/files-and-directories/file-detail/file-detail-utils"
import {FileDetail, FileDetailProps} from "../file-detail"
import * as useFileDetail from "../hooks/use-file-detail"
import {UseFileDetailHook} from "../hooks/use-file-detail"

const hookValuesDefault: UseFileDetailHook = {
  updateFile: jest.fn(),
  deselectFile: jest.fn(),
  filePreviewOption: Option.none(),
  setFilePreviewOption: jest.fn(),
  isMoveOverlayVisible: false,
  showMoveOverlay: jest.fn(),
  hideMoveOverlay: jest.fn(),
  allDirectories: directoriesForScenarioMock,
  allFiles: filesMock,
  dataLoading: false,
  isSampleCompanyPublished: false,
  isScenarioPublishedOrFinalized: false
}

const defaultProps: FileDetailProps = {
  disabled: true,
  file: fileMock,
  parentDirectory: Option.none(),
  sampleCompanyId: sampleCompaniesMock[0].id,
  isIntroFile: false,
  isLogoFile: false
}

const stateSpy = jest.spyOn(useFileDetail, "useFileDetail")

const getComponent = (props?: Partial<FileDetailProps>) => (
  <MockedProvider
    mocks={[
      {
        request: {
          query: deleteFileMutation,
          variables: {id: fileMock.id}
        },
        result: {
          data: {
            deleteFile: fileMock.id
          }
        }
      }
    ]}>
    <FileDetail {...{...defaultProps, ...props}} />
  </MockedProvider>
)

describe("file-detail", () => {
  it("renders correctly", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)
    const component = create(getComponent())
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("has correct structure", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)

    const component = mount(getComponent())
    expect(component.find(Card)).toHaveLength(1)
    expect(component.find(CardContent)).toHaveLength(1)
    expect(component.find(DetailHeader)).toHaveLength(1)
    expect(component.find(MediaFilePreview)).toHaveLength(1)
    expect(component.find(FormFieldLabel)).toHaveLength(2)
    expect(component.find(InlineEditableHeaderContainer)).toHaveLength(1)
    expect(component.find(HintText)).toHaveLength(0)
  })

  it("has correct structure intro file", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)

    const component = mount(getComponent({isIntroFile: true}))
    expect(component.find(Card)).toHaveLength(1)
    expect(component.find(CardContent)).toHaveLength(1)
    expect(component.find(DetailHeader)).toHaveLength(1)
    expect(component.find(MediaFilePreview)).toHaveLength(1)
    expect(component.find(FormFieldLabel)).toHaveLength(2)
    expect(component.find(InlineEditableHeaderContainer)).toHaveLength(1)
    expect(component.find(HintText)).toHaveLength(1)
  })

  it("has correct structure logo file", async () => {
    stateSpy.mockReturnValue(hookValuesDefault)

    const component = mount(getComponent({isLogoFile: true}))
    expect(component.find(Card)).toHaveLength(1)
    expect(component.find(CardContent)).toHaveLength(1)
    expect(component.find(DetailHeader)).toHaveLength(1)
    expect(component.find(MediaFilePreview)).toHaveLength(1)
    expect(component.find(FormFieldLabel)).toHaveLength(2)
    expect(component.find(InlineEditableHeaderContainer)).toHaveLength(1)
    expect(component.find(HintText)).toHaveLength(1)
  })

  it("renders no viewer tool if no preview is active", async () => {
    stateSpy.mockReturnValue({...hookValuesDefault, filePreviewOption: Option.none()})

    const component = mount(getComponent())
    expect(component.find(Overlay)).toHaveLength(0)
    expect(component.find(ImageViewer)).toHaveLength(0)
    expect(component.find(PdfViewer)).toHaveLength(0)
    expect(component.find(Text)).toHaveLength(2)
  })

  it("renders viewer tool if preview is active", async () => {
    stateSpy.mockReturnValue({
      ...hookValuesDefault,
      filePreviewOption: Option.of<PreviewInfo>({
        file: fileMock,
        viewerTool: ViewerToolsType.Image
      })
    })

    const component = mount(getComponent())
    expect(component.find(Overlay)).toHaveLength(1)
    expect(component.find(ImageViewer)).toHaveLength(1)
    expect(component.find(PdfViewer)).toHaveLength(0)
    expect(component.find(Text)).toHaveLength(4)
  })

  it("triggers update name correctly", async () => {
    const updateFile = jest.fn()

    stateSpy.mockReturnValue({...hookValuesDefault, updateFile})

    const component = mount(getComponent({disabled: false}))
    component.find(InlineEditableHeaderContainer).props().onConfirm("test-file")
    expect(updateFile).toHaveBeenCalledTimes(1)
    expect(updateFile).toHaveBeenCalledWith({name: "test-file"}, fileMock)
  })
})
