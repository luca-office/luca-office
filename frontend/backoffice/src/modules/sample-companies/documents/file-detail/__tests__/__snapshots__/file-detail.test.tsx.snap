// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file-detail renders correctly 1`] = `
<div
  className="css-63bajj-Card"
>
  <div
    className="card-header css-13nbzsf-CardHeader"
  >
    <div
      className="css-17beenm-DetailHeader"
    >
      <div
        className="css-1lrzij5-Icon"
        dangerouslySetInnerHTML={
          {
            "__html": {},
          }
        }
      />
      <div
        className="css-1i6jmor-Heading"
      >
        file.png
      </div>
    </div>
    <div
      className="css-ntx7po-DetailHeader"
    >
      <div
        className="css-0"
      >
        <button
          className="css-hnekdl-Button"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            <div
              className="css-mj8vo1-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </button>
      </div>
    </div>
  </div>
  <div
    className="css-1sn20oo-FileDetail"
  >
    <div
      className="card-content css-enwcrq-CardContent"
    >
      <div
        className="css-6b3ki2-FileDetail"
      >
        <div>
          <div
            className="css-13s2sfk-Heading"
          >
            files_and_directories__file_detail_name
          </div>
          <div
            className="css-0"
            id="tooltip_42"
          >
            <div
              className="css-ryq9y7-InlineEditableHeader"
            >
              <div>
                <form
                  className="css-1c5ej61-InlineEditableHeader"
                  onBlur={[Function]}
                  onSubmit={[Function]}
                >
                  <div
                    className="css-1boxw9s-InlineEditableHeader"
                  >
                    <div
                      className="css-1iwo50k"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <div
                        className="css-nrul0i"
                      >
                        <input
                          className="css-2j0qcg"
                          defaultValue="file"
                          disabled={false}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          readOnly={true}
                          type="text"
                        />
                      </div>
                    </div>
                    <div
                      className="hover-icon css-16vb3lw-Icon"
                      dangerouslySetInnerHTML={
                        {
                          "__html": {},
                        }
                      }
                    />
                  </div>
                </form>
              </div>
            </div>
          </div>
          <div
            className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje css-10zyuev-Tooltip"
            role="tooltip"
            style={
              {
                "backgroundColor": "#2e3032",
                "zIndex": 1,
              }
            }
          >
            sample_companies_files__disabled_tooltip
            <div
              className="react-tooltip-arrow styles-module_arrow__K0L3T"
              style={{}}
            />
          </div>
        </div>
        <div>
          <div
            className="css-13s2sfk-Heading"
          >
            files_and_directories__move_directory_label
          </div>
          <div
            className="css-1m3vlms-ReadonlyActionField"
            data-testid="readonly-action-field"
          >
            <div
              className="css-1bkess4-ReadonlyActionField"
            />
            <div
              className="content-wrapper css-b45n67-ReadonlyActionField"
            >
              <div
                className="content css-1147qp7-ReadonlyActionField"
              >
                <div
                  className="css-0"
                  id="tooltip_42"
                >
                  <div
                    className="css-1czgmen-PathActionField"
                  >
                    <div
                      className="css-1h826ga-PathActionField"
                    >
                      <div
                        className="css-1yjyjy3-Icon"
                        dangerouslySetInnerHTML={
                          {
                            "__html": {},
                          }
                        }
                      />
                      <div
                        className="css-1sbf56p-Text"
                      >
                        files_and_directories__move_directory_main_directory
                      </div>
                    </div>
                    <div
                      className="css-1a85mc8-Text"
                      onClick={[Function]}
                    >
                      files_and_directories__move_directory
                    </div>
                  </div>
                </div>
                <div
                  className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje css-10zyuev-Tooltip"
                  role="tooltip"
                  style={
                    {
                      "backgroundColor": "#2e3032",
                      "zIndex": 1,
                    }
                  }
                >
                  sample_companies_files__disabled_tooltip
                  <div
                    className="react-tooltip-arrow styles-module_arrow__K0L3T"
                    style={{}}
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="css-1kpejq0-MediaFilePreview"
      >
        <div
          className="css-g5r2b1-MediaFilePreview"
        >
          <div
            className="css-wc80za-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
          <button
            className="css-jeltaf-Button"
            disabled={false}
            onClick={[Function]}
          >
            <div
              className="css-95j2hm-Button"
            >
              <div
                className="css-kuxry5-Icon"
                dangerouslySetInnerHTML={
                  {
                    "__html": {},
                  }
                }
              />
              files_and_directories__open_in_viewer
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
